subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/TestReport.java

```diff
@@ -64,26 +68,63 @@ protected ObjectFactory getObjectFactory() {
     /**
      * Returns the directory to write the HTML report to.
      */
-    @OutputDirectory
+    @Deprecated
+    @ReplacedBy("destinationDirectory")
     public File getDestinationDir() {
-        return destinationDir;
+        DeprecationLogger.deprecateProperty(TestReport.class, "destinationDir").replaceWith("destinationDirectory")
+            .willBeRemovedInGradle8()
+            .withDslReference()
+            .nagUser();
+        return getDestinationDirectory().get().getAsFile();
     }
 
     /**
      * Sets the directory to write the HTML report to.
      */
+    @Deprecated
     public void setDestinationDir(File destinationDir) {
-        this.destinationDir = destinationDir;
+        DeprecationLogger.deprecateProperty(TestReport.class, "destinationDir").replaceWith("destinationDirectory")
+            .willBeRemovedInGradle8()
+            .withDslReference()
+            .nagUser();
+        getDestinationDirectory().set(destinationDir);
+    }
+
+    /**
+     * Returns the directory to write the HTML report to.
+     *
+     * @since 7.4
+     */
+    @Incubating
+    @OutputDirectory
+    public DirectoryProperty getDestinationDirectory() {
+        return this.destinationDir;
     }
 
     /**
      * Returns the set of binary test results to include in the report.
      */
+    @Deprecated
+    @ReplacedBy("testResults")
```

Comment:
Does this need to be marked `@Internal`?

Reply:
I'm not sure. I did get validation errors when mixing `@ReplacedBy` and `@InputFile`, for example.

Reply:
I confirmed that `@ReplacedBy` and `@Internal` cannot coexist:

```
  - Type 'org.gradle.api.tasks.testing.TestReport' property 'testResultDirs' annotated with @ReplacedBy should not be also annotated with @Internal.
    
    Reason: A property is ignored but also has input annotations.
    
    Possible solutions:
      1. Remove the input annotations.
      2. Remove the @ReplacedBy annotation.
```

Reply:
ah ok, I forgot we had done that.